app/models/spree/order.rb

```diff
@@ -433,6 +433,7 @@ def empty!
       all_adjustments.destroy_all
       payments.clear
       shipments.destroy_all
+      restart_checkout_flow if state == "payment"
```

Comment:
Have not tested yet - will stage deploy right after this comment - but this makes sense to me :+1: 

Also, we'll probably need the same approach to address #9131 , but with state confirmation in mind. Anyway, something for a separate PR.